sdks/java/core/src/main/java/org/apache/beam/sdk/util/SqlRows.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.util;
+
+import static org.apache.beam.sdk.util.Preconditions.checkArgumentNotNull;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.schemas.SchemaProvider;
+import org.apache.beam.sdk.transforms.ProcessFunction;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.SimpleFunction;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.sdk.values.TypeDescriptor;
+
+public final class SqlRows {
```

Comment:
done

Reply:
It's just a tiny step towards whatever abstractions we want to end up with, as it doesn't even cover Avro yet.
I'm still trying to wrap my head around how would a proper data format abstraction look like in Beam.
I mean, instead of `FileIO`, `AvroIO`, `KafkaIO`, `ThriftIO` plus Kafka tables for each data format, what if we took the format out of the IO as an entirely different concern and provide the ability to mix those freely. If IOs are just `[File, Kafka, PubSub, Socket, whatever]`, and data formats are `[Avro, Protobuf, CSV, Thrift, whatever]`, we can (in theory) allow any combination like PubSub+Thrift or File+Proto with no extra cost (except for incompatible candidates, like Kafka+parquet, perhaps).

Do you think this is worth considering (for 3.0, of course)? I am aware that such an abstraction is really difficult to get right (if at all feasible), but I still think it may be worth trying, as we should at least end up reducing the number of data_source+data_format combinations that are now modelled as new datasource types.

Reply:
Yes this sounds ideal. As far as 3.0 - I think it would make sense to start working on it in 2.x versions as an `@Experimental` API. We shouldn't remove the existing IOs prior to a major version change, but it would be fine to start work on an alternative. This would be similar to (and related to) the `SchemaIOProvider` concept we had a Google intern work on last summer:
- https://s.apache.org/schemaio-development-guide
- http://s.apache.org/beam-schema-io

Would you have any interest in working on an effort like this?

CC: @robinyqiu a design like this would be relevant for your work on IOs

Reply:
I'd be more than happy to get involved. Most likely, I'll need to join your slack channel, and for that I need an apache.org email, apparently. I'll see how I can sort that out...

Reply:
It looks like you can get an invite at https://s.apache.org/slack-invite (more info here: https://infra.apache.org/slack.html)

Reply:
Sounds easy. Not sure who's supposed to invite me, but I'll ask around. If you can help with that, I'd be grateful.

Reply:
Sure I can invite you as a guest. I think https://s.apache.org/slack-invite is _supposed_ to be self-service but the page indicates it breaks it often. Can you send me a message at bhulette@apache.org so I know what email to use?